rpcs3/Emu/Cell/lv2/sys_usbd.cpp

```diff
@@ -282,14 +336,122 @@ void usb_handler_thread::operator()()
+		// If there are no handled devices the usb thread is not actively needed, so we can sleep here for a while
 		if (handled_devices.empty())
 			std::this_thread::sleep_for(500ms);
 		else
 			std::this_thread::sleep_for(200us);
 	}
 }
 
+void usb_handler_thread::perform_device_scan()
+{
+	std::unordered_set<libusb_device_descriptor> detach_scan;
+	for (auto&& a : device_scan_list)
+		detach_scan.emplace(a);
+
+	// look if any device which we could be interested in is actually connected
+	libusb_device** list = nullptr;
+	ssize_t ndev = libusb_get_device_list(ctx, &list);
+	for (ssize_t index = 0; index < ndev; index++)
+	{
+		libusb_device_descriptor desc;
+		int rc = libusb_get_device_descriptor(list[index], &desc);
+		if (rc != LIBUSB_SUCCESS)
+		{
+			sys_usbd.error("%s", libusb_strerror(rc));
+			continue;
+		}
+		if (!device_scan_list.contains(desc))
+		{
+			for (const auto& [id_vendor, id_product_min, id_product_max, s_name] : device_whitelist)
+			{
+				// attach
+				if (desc.idVendor == id_vendor && desc.idProduct >= id_product_min && desc.idProduct <= id_product_max)
+				{
+					sys_usbd.success("Found device: %s", s_name);
+					usb_devices.emplace_back(std::make_shared<usb_device_passthrough>(s_name, list[index], desc));
+					advise_ldds_of_attach(*usb_devices.back().get());
+				}
+			}
+			device_scan_list.emplace(desc);
+		}
+
+		if (detach_scan.contains(desc))
+			detach_scan.erase(desc);
+	}
+
+	for (auto& item : detach_scan)
+	{
+		// TODO: detach
+		for (auto dev = usb_devices.begin(); dev != usb_devices.end(); )
+		{
+			auto usb_dev = dynamic_cast<usb_device_passthrough*>(dev->get());
+			if (!usb_dev)
+				continue;
+
+			libusb_device_descriptor desc;
+			usb_dev->get_descriptor(&desc);
+
+			if (desc.idVendor == item.idVendor
+				&& desc.idProduct == item.idProduct
+				&& desc.iSerialNumber == item.iSerialNumber)
+			{
+				sys_usbd.success("Removing device: %s", usb_dev->get_name());
+				advise_ldds_of_detach(*dev->get());
+				dev = usb_devices.erase(dev);
+			}
+			else
+				dev++;
+		}
+
+		device_scan_list.erase(item);
+	}
+
+	libusb_free_device_list(list, 1);
+}
+
+int LIBUSB_CALL usb_handler_thread::hotplug_callback(struct libusb_context* ctx, struct libusb_device* dev, libusb_hotplug_event event, void* thread_ref)
+{
+	// since the hotplug callback was received from a Cstyle callback to the static function, we will proxy
+	// it back to the member function on the thread that created the callback.
+	auto thread          = reinterpret_cast<usb_handler_thread*>(thread_ref);
+	auto instance_mem_fn = std::mem_fn(&usb_handler_thread::hotplug_callback_mem_fn);
+	return instance_mem_fn(thread, ctx, dev, event);
+};
+
+int usb_handler_thread::hotplug_callback_mem_fn(struct libusb_context* ctx, struct libusb_device* dev, libusb_hotplug_event event)
+{
+	static libusb_device_handle* dev_handle = NULL;
+	struct libusb_device_descriptor desc;
+	int rc;
+
+	(void)libusb_get_device_descriptor(dev, &desc);
```

Comment:
Why void

Reply:
Copy paste from libusb... 'fixed' in next push